heracles/dices/utils.py

```diff
@@ -0,0 +1,190 @@
+    return _Cls
+
+
+def sub_to_Cls(Cls, x):
+    """
+    Substracts a dictionary of Cl values to another.
+    input:
+        Cls: dictionary of Cl values
+        x: dictionary of Cl values
+    returns:
+        Cls: updated dictionary of Cl values
+    """
+    _Cls = {}
+    for key in Cls.keys():
+        ell = Cls[key].ell
+        _Cls[key] = Result(Cls[key].__array__() - x[key], ell)
+    return _Cls
+
+
+# Generic covariance calculation
+def cov2corr(cov):
+    """
+    Produces a correlation matrix from a covariance matrix.
+    input:
+        cov: covariance matrix
+    returns:
+        corr: correlation matrix
+    """
+    corr = np.copy(cov)
+    sig = np.sqrt(np.diag(cov))
+    corr /= np.outer(sig, sig)
+    return corr
+
+
+def make_posdef(matrix):
+    """
+    Ensure the given symmetric matrix is positive definite.
+    If not, adjust eigenvalues to make it positive definite.
+    """
+    if not np.allclose(matrix, matrix.T):
+        print("Matrix is not symmetric.")
+        matrix = 0.5 * (matrix + matrix.T)
+    v, w = np.linalg.eigh(matrix)
+    v[v < 0.0] = np.min(np.abs(v))
+    return w @ np.diag(v) @ w.T
+
+
+# Summary statistics for Cls
+def get_mean_Cljk(Clss):
+    Cl_mu = {}
+    total = len(Clss)
+    for i, key in enumerate(list(Clss.keys())):
+        Cls = Clss[key]
+        if i == 0:
+            for key in Cls.keys():
+                cl = Cls[key]
+                ell = cl.ell
+                Cl_mu[key] = Result(cl, ell)
+        else:
+            for key in Cls.keys():
+                cl = Cls[key].__array__() + Cl_mu[key].__array__()
+                ell = Cls[key].ell
+                Cl_mu[key] = Result(cl, ell)
+    for key in Cl_mu.keys():
+        cl = Cl_mu[key]
+        ell = cl.ell
+        Cl_mu[key] = Result(cl.__array__() / total, ell)
+    return Cl_mu
+
+
+def get_W(x, xbar, jk=False):
+    """
+    Internal method to compute the W matrices.
+    input:
+        x: Cl
+        xbar: mean Cl
+        jk: if True, computes the jackknife version of the W matrices
+    returns:
+        W: W matrices
+    """
+    W = []
+    _xbi, _xbj = np.meshgrid(xbar, xbar, indexing="ij")
+    for i in range(0, len(x)):
+        _xi, _xj = np.meshgrid(x[i], x[i], indexing="ij")
+        _Wk = (_xi - _xbi) * (_xj - _xbj)
+        W.append(_Wk)
+    W = np.array(W)
+    if jk:
+        n = len(x)
+        W *= ((n - 1) ** 2.0) / n
+    return W
+
+
+def get_Wbar(x, xbar):
+    """
+    Internal method to compute the W matrices.
+    input:
+        x: Cl
+        xbar: mean Cl
+        jk: if True, computes the jackknife version of the W matrices
+    returns:
+        W: W matrices
+    """
+    W = np.zeros((len(xbar), len(xbar)))
+    n = len(x)
+    _xbi, _xbj = np.meshgrid(xbar, xbar, indexing="ij")
+    for i in range(0, len(x)):
+        _xi, _xj = np.meshgrid(x[i], x[i], indexing="ij")
+        W += (_xi - _xbi) * (_xj - _xbj)
+    W *= 1/n
+    return W
```

Comment:
Ok, thanks to all of the refactoring, this functions is clearer now: it computes the sample covariance! We can use a one-pass algorithm that doesn't need pre-computation of `xbar`, is numerically stable, and works on inputs of any shape:

```py
def sample_covariance(samples):
    """
    Returns the sample covariance matrix of *samples*.  The first axis
    is taken as different realisations.  Works on multi-dimensional
    inputs.
    """
    n, *dim = samples.shape
    mu = np.zeros((*dim,))
    cov = np.zeros((*dim, *dim))
    for i in range(n):
        x = samples[i]
        delta = x - mu
        mu += delta / (i + 1)
        cov += (np.multiply.outer(delta, x - mu) - cov) / (i + 1)
    return cov
```

Reply:
Implemented!